Invalid macros for SchemaFor with Record having Option of Record with ADT

```
import com.sksamuel.avro4s.SchemaFor

sealed trait Adt
case object One extends Adt
case object Two extends Adt

case class Example(adt: Adt)
case class HasOption(opt: Option[Example])

SchemaFor[HasOption]
```
causes
```
java.lang.NoClassDefFoundError: $anon$1/_schema $$anon
  ... 30 elided
Caused by: java.lang.ClassNotFoundException: $anon$1._schema $$anon
  at scala.reflect.internal.util.AbstractFileClassLoader.findClass(AbstractFileClassLoader.scala:72)
  at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
  at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
  ... 30 more
```

But the following works (we keep same derivation rules, but factor out part of the macro into companion object) :
```
import com.sksamuel.avro4s.SchemaFor

sealed trait Adt
case object One extends Adt
case object Two extends Adt

case class Example(adt: Adt); object Example {
  implicit val schemaFor: SchemaFor[Example] = com.sksamuel.avro4s.SchemaFor.applyMacro
}
case class HasOption(opt: Option[Example])

SchemaFor[HasOption]
```
outputs `com.sksamuel.avro4s.SchemaFor[HasOption] = $anon$1@6165ac45`

--- a/avro4s-core/src/test/resources/record_option_record_adt.json
+++ b/avro4s-core/src/test/resources/record_option_record_adt.json
@@ -1,0 +1,22 @@
+{
+  "type" : "record",
+  "name" : "HasOption",
+  "namespace" : "com.sksamuel.avro4s.schema",
+  "fields" : [ {
+    "name" : "option",
+    "type" : [ "null", {
+      "type" : "record",
+      "name" : "HasAdt",
+      "namespace" : "com.sksamuel.avro4s.schema",
+      "fields" : [ {
+        "name" : "adt",
+        "type" : {
+          "type" : "enum",
+          "name" : "Adt",
+          "namespace" : "com.sksamuel.avro4s.schema",
+          "symbols" : [ "One", "Two" ]
+        }
+      } ]
+    } ]
+  } ]
+}

--- a/avro4s-core/src/test/scala/com/sksamuel/avro4s/schema/RecordOptionRecordAdt.scala
+++ b/avro4s-core/src/test/scala/com/sksamuel/avro4s/schema/RecordOptionRecordAdt.scala
@@ -1,0 +1,22 @@
+package com.sksamuel.avro4s.schema
+
+import com.sksamuel.avro4s.AvroSchema
+import org.scalatest.{FunSuite, Matchers}
+
+// https://github.com/sksamuel/avro4s/issues/334
+class RecordOptionRecordAdt extends FunSuite with Matchers {
+
+  test("records with optional records with adt"){
+    val schema = AvroSchema[HasOption]
+    val expected = new org.apache.avro.Schema.Parser().parse(getClass.getResourceAsStream("/record_option_record_adt.json"))
+    schema.toString(true) shouldBe expected.toString(true)
+  }
+}
+
+sealed trait Adt
+case object One extends Adt
+case object Two extends Adt
+
+case class HasAdt(adt: Adt)
+
+case class HasOption(option: Option[HasAdt])

--- a/avro4s-macros/src/main/scala/com/sksamuel/avro4s/SchemaFor.scala
+++ b/avro4s-macros/src/main/scala/com/sksamuel/avro4s/SchemaFor.scala
@@ -299,12 +299,12 @@ object SchemaFor extends TupleSchemaFor with CoproductSchemaFor {
       val annos = reflect.annotationsqq(tpe.typeSymbol)
 
       c.Expr[SchemaFor[T]](
-        q"""
-        new _root_.com.sksamuel.avro4s.SchemaFor[$tpe] {
-          private val _schema = _root_.com.sksamuel.avro4s.SchemaFor.buildSchema($name, $namespace, Seq(..$fields), Seq(..$annos), $isValueClass)
-          override def schema: _root_.org.apache.avro.Schema = _schema
-        }
-      """)
+        q"""{
+          val _schema = _root_.com.sksamuel.avro4s.SchemaFor.buildSchema($name, $namespace, Seq(..$fields), Seq(..$annos), $isValueClass)
+          new _root_.com.sksamuel.avro4s.SchemaFor[$tpe] {
+            override def schema: _root_.org.apache.avro.Schema = _schema
+          }
+        }""")
     }
   }
 

